### PR TITLE
replica, sstable: replace generation_type::value() with generation_type::as_int()

### PR DESCRIPTION
**Pull request by tchaikov**
this series prepares for the UUID based generation by replacing the general `value()` function with the function with more specific name: `as_int()`.

### CODE REVIEW

**Comment by tchaikov** [comment]
this change might conflict with #13782. these two changeset are posted separately for two reasons

1. to get them reviewed sooner
2. despite that they serve the same purpose -- to prepare for the UUID based generation, they accomplish different smaller goals.

**Comment by tchaikov** [comment]
Can we implement this in smaller steps?

**Comment by denesb** [comment]
> Can we implement this in smaller steps?

The thing is, as the code currently is, adding support for UUID generation type will require changes at each and every site this PR touches. Doing what I suggested, would mean that we could seamlessly switch and those code sites wouldn't even notice.

**Comment by tchaikov** [comment]
yeah, i understand the merit of your proposal. that's actually part of my plan. but i really want to implement things in smaller steps. as long as a change is self-contained, and it is aligned with the direction in general. probably we can land it and move on with the next step? to "perform the seemless switch" is a much bigger change than this one. and we have quite a lot of caller sites where we are still using integer. i have a patch which only implements part of this switch. i don't dare to change *all* of them. and i'd prefer leave it do another PR.

**Comment by denesb** [comment]
I just didn't see how this PR takes us closer to an easier switch, but if you think it has value, I'm fine with merging it.

**Comment by tchaikov** [comment]
> I just didn't see how this PR takes us closer to an easier switch, but if you think it has value, I'm fine with merging it.

this change was actually extracted from a larger patch which performs the switch you mentioned. the value of it is: it renames `value()`. so the `to_int()` calls stands out. some of the `to_int()` calls will stay with us much longer. they are the ones in tests. and some of them will be absorbed into an other method `owned_by_this_shard()`.

**Comment by raphaelsc** [comment]
> > I just didn't see how this PR takes us closer to an easier switch, but if you think it has value, I'm fine with merging it.
> 
> this change was actually extracted from a larger patch which performs the switch you mentioned. the value of it is: it renames `value()`. so the `to_int()` calls stand out. some of the `to_int()` calls will stay with us much longer. they are the ones in tests. and some of them will be absorbed into an other method `owned_by_this_shard()`.

By the way, I'd only like to clarify the sstable generation id is never an authority of shard ownership for a sstable. It only provides a hint, so the name have to make it explicit. e.g.: shard_id generation_type::hint_on_shard_ownership();

**Comment by tchaikov** [comment]
okay, how about `maybe_owned_by_this_shard()` ? as hints are very general. it reminds me of `fadvise(2)` =)

File: replica/table.cc [hunk 1]
```diff
@@ -67,7 +67,7 @@ static seastar::metrics::label keyspace_label("ks");
 using namespace std::chrono_literals;
 
 void table::update_sstables_known_generation(std::optional<sstables::generation_type> generation) {
-    auto gen = generation.value_or(sstables::generation_type(0)).value();
+    auto gen = generation.value_or(sstables::generation_type(0)).as_int();
     if (_sstable_generation_generator) {
```
Comment (raphaelsc) [inline replica/table.cc:71]: By the way, we can easily get rid of generation_type::int_t usage here. we can feed generation_type directly into the generator.

File: sstables/sstable_directory.cc [hunk 2]
```diff
@@ -293,7 +293,7 @@ future<> sstable_directory::system_keyspace_components_lister::process(sstable_d
             // FIXME -- handle
             return make_ready_future<>();
         }
-        if ((generation_value(desc.generation) % smp::count) != this_shard_id()) {
+        if ((desc.generation.as_int() % smp::count) != this_shard_id()) {
             return make_ready_future<>();
```
Comment (raphaelsc) [inline sstables/sstable_directory.cc:297]: maybe_owned_by_this_shard() will fit greatly here :-)

File: sstables/sstable_directory.cc [hunk 3]
```diff
@@ -573,7 +573,7 @@ highest_generation_seen(sharded<sstables::sstable_directory>& directory) {
             return sstables::generation_type(0);
         }
     });
-    co_return highest.value() ? std::make_optional(highest): std::nullopt;
+    co_return highest.as_int() ? std::make_optional(highest) : std::nullopt;
 }
```
Comment (raphaelsc) [inline sstables/sstable_directory.cc:577]: generation_type::uninitialized()

File: sstables/sstable_directory.cc (hunk 2 above)
Comment (tchaikov) [reply sstables/sstable_directory.cc:297]: #13846

**Comment by tchaikov** [comment]
> > I just didn't see how this PR takes us closer to an easier switch, but if you think it has value, I'm fine with merging it.
> 
> this change was actually extracted from a larger patch which performs the switch you mentioned. the value of it is: it renames `value()`. so the `to_int()` calls stand out. some of the `to_int()` calls will stay with us much longer. they are the ones in tests. and some of them will be absorbed into an other method `owned_by_this_shard()`.

some of the tests are now using `generation_type`, see #13845

**Pull request merged**